# Step 1: Data Cleaning and Feature Matrix Construction
This notebook should run the authoritative v3 Step 1 pipeline for the GAITNDD dataset.

**What Step 1 does:**
Loads the 64 raw `.ts` files from PhysioNet GAITNDD v1.0.0, applies the v3 hard plausibility filter plus the per-subject robust outlier filter, engineers the 14-feature v3 matrix, and writes the shared coordination artifacts used by the later notebooks.

**Authoritative v3 preprocessing targets:**
1. Hard plausibility filter: remove physiologically impossible rows, including the negative `double_support_*` artifacts and stride durations above 3.0 s.
2. Robust subject-wise outlier filter: remove turn/pause contamination on `left_stride_s`, `right_stride_s`, and `double_support_s` when enough clean strides remain.
3. Feature construction: keep the raw timing block, remove `left_stance_pct`, `right_stance_pct`, and `stride_asymmetry_signed`, and recompute `asymmetry_index`, `cv_stride`, `cv_swing`, and `dfa_alpha_stride` after cleaning.

**Expected v3 outputs after rerun:**
- `v3/gait_features_v3.csv`: 13,765 rows × 17 columns (14 features + `subject_id`, `condition`, `label`)
- `v3/control_partition_v3.json`: balanced 8/8 control split with age delta `0.125` years and gait-speed delta `0.00375 m/s`
- `v3/preprocessing_manifest_v3.json`: filter statistics and control-partition summary for auditability

The checked-in v3 manifest reports `409` rows removed by the hard filter and `986` additional rows removed by the robust filter. If this notebook still shows the old 20-column output after opening, rerun Cell 2: the code below has now been updated to call the explicit v3 pipeline rather than the default v2 path.


In [8]:
import sys
import os
from pathlib import Path

sys.path.insert(0, str(Path('..') / 'src'))

import polars as pl
from features import build_feature_matrix, get_feature_cols

FEATURE_COLS = get_feature_cols('v3')

DATA_DIR = '../data/raw/gait-in-neurodegenerative-disease-database-1.0.0'
PROCESSED_DIR = '../data/processed'
OUTPUT_FILENAME = 'v3/gait_features_v3.csv'


In [9]:
# Run the full Step 1 pipeline.
# Loads raw data, applies the v3 artifact filters, engineers the 14-feature v3 set,
# saves v3/gait_features_v3.csv and v3/control_partition_v3.json, and returns the DataFrame.
df, part = build_feature_matrix(
    DATA_DIR,
    PROCESSED_DIR,
    output_filename=OUTPUT_FILENAME,
    feature_cols=FEATURE_COLS,
    feature_set_version='v3',
    filter_strategy='v3',
    control_partition_version='v3',
    partition_output_filename='v3/control_partition_v3.json',
    manifest_filename='v3/preprocessing_manifest_v3.json',
)

print(f'Feature matrix shape: {df.shape}')
print(f'Columns ({len(df.columns)}): {df.columns}')


Feature matrix shape: (13765, 17)
Columns (17): ['left_stride_s', 'right_stride_s', 'left_swing_s', 'right_swing_s', 'left_swing_pct', 'right_swing_pct', 'left_stance_s', 'right_stance_s', 'double_support_s', 'double_support_pct', 'asymmetry_index', 'cv_stride', 'cv_swing', 'dfa_alpha_stride', 'subject_id', 'condition', 'label']


In [10]:
counts = (
    df
    .group_by('condition')
    .agg(
        pl.n_unique('subject_id').alias('subjects'),
        pl.len().alias('strides'),
    )
    .sort('condition')
)
print('Subject and stride counts per condition:')
print(counts)

Subject and stride counts per condition:
shape: (4, 3)
┌───────────┬──────────┬─────────┐
│ condition ┆ subjects ┆ strides │
│ ---       ┆ ---      ┆ ---     │
│ str       ┆ u32      ┆ u32     │
╞═══════════╪══════════╪═════════╡
│ als       ┆ 13       ┆ 2285    │
│ control   ┆ 16       ┆ 3786    │
│ hd        ┆ 19       ┆ 4280    │
│ pd        ┆ 15       ┆ 3414    │
└───────────┴──────────┴─────────┘


In [11]:
ai = df['asymmetry_index']
print('asymmetry_index (per-stride):')
print(f'  min  = {ai.min():.6f}')
print(f'  max  = {ai.max():.6f}')
print(f'  mean = {ai.mean():.6f}')

asymmetry_index (per-stride):
  min  = 0.000000
  max  = 0.660284
  mean = 0.031952


In [12]:
# cv_stride is a subject-level property broadcast to all stride rows,
# so deduplicate by subject before computing statistics.
cv_per_subject = (
    df
    .unique(subset=['subject_id'])
    .select(['subject_id', 'cv_stride'])
    .sort('subject_id')
)
cv_vals = cv_per_subject['cv_stride']
print(f'cv_stride (per-subject, n={len(cv_per_subject)}):')
print(f'  min  = {cv_vals.min():.6f}')
print(f'  max  = {cv_vals.max():.6f}')
print(f'  mean = {cv_vals.mean():.6f}')

cv_stride (per-subject, n=63):
  min  = 0.011540
  max  = 0.201089
  mean = 0.050133


In [13]:
print('First 5 rows:')
print(df.head(5))

First 5 rows:
shape: (5, 17)
┌────────────┬────────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────┐
│ left_strid ┆ right_stri ┆ left_swing ┆ right_swi ┆ … ┆ dfa_alpha ┆ subject_i ┆ condition ┆ label │
│ e_s        ┆ de_s       ┆ _s         ┆ ng_s      ┆   ┆ _stride   ┆ d         ┆ ---       ┆ ---   │
│ ---        ┆ ---        ┆ ---        ┆ ---       ┆   ┆ ---       ┆ ---       ┆ str       ┆ i8    │
│ f64        ┆ f64        ┆ f64        ┆ f64       ┆   ┆ f64       ┆ str       ┆           ┆       │
╞════════════╪════════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════╡
│ 1.2833     ┆ 1.3533     ┆ 0.4067     ┆ 0.4133    ┆ … ┆ 1.114817  ┆ als1      ┆ als       ┆ 1     │
│ 1.3233     ┆ 1.2667     ┆ 0.4833     ┆ 0.4       ┆ … ┆ 1.114817  ┆ als1      ┆ als       ┆ 1     │
│ 1.3033     ┆ 1.36       ┆ 0.45       ┆ 0.4267    ┆ … ┆ 1.114817  ┆ als1      ┆ als       ┆ 1     │
│ 1.4167     ┆ 1.2833     ┆ 0.5033     ┆ 0.3667    ┆ … ┆ 1.114

In [14]:
csv_path = Path(PROCESSED_DIR) / 'v3' / 'gait_features_v3.csv'
json_path = Path(PROCESSED_DIR) / 'v3/control_partition_v3.json'

print(f'v3/gait_features_v3.csv exists: {csv_path.exists()}')
print(f'v3/control_partition_v3.json exists: {json_path.exists()}')

# Reload CSV and confirm shape matches the in-memory DataFrame.
reloaded = pl.read_csv(str(csv_path))
print(f'Reloaded CSV shape: {reloaded.shape}  (matches in-memory: {reloaded.shape == df.shape})')

# Confirm no nulls in any feature column.
null_total = df.select(FEATURE_COLS).null_count().to_numpy().sum()
print(f'Null values in feature columns: {null_total}')


v3/gait_features_v3.csv exists: True
v3/control_partition_v3.json exists: True
Reloaded CSV shape: (13765, 17)  (matches in-memory: True)
Null values in feature columns: 0


## Step 1 Verification Summary
| Checkpoint | Expected v3 value | Status after rerun |
|---|---:|---|
| Raw strides | 15,160 | from raw PhysioNet files |
| Removed by hard plausibility filter | 409 | should match manifest |
| Removed by robust outlier filter | 986 | should match manifest |
| Final clean strides | 13,765 | should match `gait_features_v3.csv` |
| PD subjects / strides | 15 / 3,414 | should match rerun output |
| HD subjects / strides | 19 / 4,280 | should match rerun output |
| ALS subjects / strides | 13 / 2,285 | should match rerun output |
| Control subjects / strides | 16 / 3,786 | should match rerun output |
| Feature columns | 14 | v3 feature list only |
| Total CSV columns | 17 | 14 features + 3 metadata columns |
| Null values in feature columns | 0 | should remain zero |
| `v3/gait_features_v3.csv` written | `True` | verify in Cell 7 |
| `v3/control_partition_v3.json` written | `True` | verify in Cell 7 |

The authoritative Step 1 coordination artifact is the robustly filtered 14-feature matrix. A rerun of Cell 2 should now produce the correct `13,765 × 17` shape rather than the earlier 20-column v2-shaped output.
